src/libraries/Common/src/System/IO/ReadOnlyMemoryStream.cs

```diff
@@ -96,23 +128,27 @@ public override Task<int> ReadAsync(byte[] buffer, int offset, int count, Cancel
             ValidateBufferArguments(buffer, offset, count);
             return cancellationToken.IsCancellationRequested ?
                 Task.FromCanceled<int>(cancellationToken) :
-                Task.FromResult(Read(new Span<byte>(buffer, offset, count)));
+                Task.FromResult(ReadBuffer(new Span<byte>(buffer, offset, count)));
```

Comment:
We should probably add a call to EnsureNotClosed before the return statement, so that it still is checked even if cancellation is already requested.

Reply:
Should this be backported?

Reply:
Both FileSteam and MemoryStream have the current behavior, I did consider this when adding the dispose checks.
https://github.com/dotnet/runtime/blob/630069daca87b7dbfaf91d0bd91a730f5218afe4/src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs#L378-L382
https://github.com/dotnet/runtime/blob/630069daca87b7dbfaf91d0bd91a730f5218afe4/src/libraries/System.Private.CoreLib/src/System/IO/MemoryStream.cs#L386-L387

Do you still think it's better to check disposed before cancelled?

Reply:
@stephentoub please let me know if you still think this should be changed, since the current behavior matches that of our other streams.

Reply:
> since the current behavior matches that of our other streams

Some of them.  Others it doesn't, e.g.

https://github.com/dotnet/runtime/blob/3c801a8a3b29eabc29ae88e1961c772fda17a1d2/src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs#L369-L378

https://github.com/dotnet/runtime/blob/3c801a8a3b29eabc29ae88e1961c772fda17a1d2/src/libraries/System.Net.Sockets/src/System/Net/Sockets/NetworkStream.cs#L516-L519

I think the right behavior is to flag misuse first.  But as you've highlighted, we're inconsistent, so go ahead with what you have for now.

Reply:
Do you think it makes sense to standardize this behavior?  Does it complicate the code that folks write by throwing here, or should the never be calling when token is cancelled and stream is disposed?  I can imagine some scenario where someone chains a ReadAsync on a continuation and doesn't check CT, but cancels and disposes on another thread.

Reply:
> should they never be calling when token is cancelled and stream is disposed?

It's fine to call when the token is canceled.  It's not fine to call when the stream is disposed, just like in general it's not fine to use a disposed object.

> Do you think it makes sense to standardize this behavior?

Yes, I think we should subsequently do a pass to check that we're validating stream state before we're checking the cancellation token.

Reply:
I see, since you plan to standardize on that behavior then I'll go ahead and change it now.  Need to reset the CI anyhow.